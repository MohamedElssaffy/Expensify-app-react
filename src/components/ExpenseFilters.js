import React from 'react';
import { DateRangePicker } from 'react-dates';
import { connect } from 'react-redux';
import {
  setEndDate,
  setStartDate,
  setText,
  sortByAmount,
  sortByDate,
} from '../actions/filters';

export class ExpenseFilters extends React.Component {
  state = {
    focusedInput: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (focusedInput) => {
    this.setState(() => ({
      focusedInput,
    }));
  };

  onTextChange = (e) => {
    this.props.setText(e.target.value);
  };

  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    }
    if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div className='container'>
        <div className='input-group'>
          <div className='input-group__item'>
            <input
              className='text-input'
              placeholder='Search expenses'
              type='text'
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className='input-group__item'>
            <select
              className='select'
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value='date'>Date</option>
              <option value='amount'>Amount</option>
            </select>
          </div>
          <div className='input-group__item'>
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.focusedInput}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

const mapDispatchToProps = (dispatch) => ({
  setText(text) {
    dispatch(setText(text));
  },
  setStartDate(startDate) {
    dispatch(setStartDate(startDate));
  },
  setEndDate(endDate) {
    dispatch(setEndDate(endDate));
  },
  sortByDate() {
    dispatch(sortByDate());
  },
  sortByAmount() {
    dispatch(sortByAmount());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseFilters);
