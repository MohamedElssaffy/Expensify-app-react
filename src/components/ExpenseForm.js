import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.description ? props.description : '',
      note: props.note ? props.note : '',
      amount: props.amount ? (props.amount / 100).toString() : '',
      createdAt: props.createdAt ? moment(props.createdAt) : moment(),
      focused: false,
      error: '',
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({
      description,
    }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({
      note,
    }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount,
      }));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ focused }));
  };

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount',
      }));
    } else {
      this.setState(() => ({ error: '' }));

      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <div>
        <h2>Add Expense</h2>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type='text'
            placeholder='Description'
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type='text'
            value={this.state.amount}
            onChange={this.onAmountChange}
            placeholder='Amount'
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.focused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            value={this.state.note}
            onChange={this.onNoteChange}
            placeholder='Add some note to remmber what you want from this expense excatly (optional)'
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
