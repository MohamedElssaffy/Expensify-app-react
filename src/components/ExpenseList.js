import React from 'react';
import { connect } from 'react-redux';

import ExpenseItem from './ExpenseItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div className='container'>
    <div className='list-header'>
      <p className='show-for-mobile'>Expenses</p>
      <p className='show-for-desktop'>Expense</p>
      <p className='show-for-desktop'>Amount</p>
    </div>
    <div className='list-body'>
      {props.expenses.length === 0 ? (
        <div className='list-item list-item--message'>
          <span>No Expenses</span>
        </div>
      ) : (
        props.expenses.map((expense) => (
          <ExpenseItem key={expense.id} {...expense} />
        ))
      )}
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state),
  };
};

export default connect(mapStateToProps)(ExpenseList);
