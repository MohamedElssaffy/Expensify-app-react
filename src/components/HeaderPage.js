import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogout } from '../actions/auth';

export const Header = (props) => (
  <div>
    <h1>Expensify</h1>
    <NavLink to='/dashboard' activeClassName='is-active'>
      Home
    </NavLink>
    <NavLink to='/create' activeClassName='is-active'>
      Create
    </NavLink>
    <button onClick={props.startLogout}>Log out</button>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout() {
    return dispatch(startLogout());
  },
});

export default connect(undefined, mapDispatchToProps)(Header);
