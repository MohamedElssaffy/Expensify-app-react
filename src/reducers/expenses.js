// Expenses Reducer

const defaultExpState = [];

export default (state = defaultExpState, action) => {
  switch (action.type) {
    case 'SET_EXPENSES':
      return action.expenses;
    case 'ADD_EXPENSE':
      return [...state, action.expense];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map((exp) => {
        if (exp.id === action.id) {
          return {
            ...exp,
            ...action.updates,
          };
        }
        return exp;
      });
    default:
      return state;
  }
};
