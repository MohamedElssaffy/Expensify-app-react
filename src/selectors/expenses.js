// get visible expenses

export default ({ expenses, filters }) => {
  const { text = '', sortBy = 'date', startDate, endDate } = filters;

  return expenses
    .filter((exp) => {
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(exp.createdAt, 'day')
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(exp.createdAt, 'day')
        : true;
      const textMatch = exp.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt > b.createdAt ? -1 : 1;
      }
      if (sortBy === 'amount') {
        return a.amount > b.amount ? -1 : 1;
      }
    });
};
