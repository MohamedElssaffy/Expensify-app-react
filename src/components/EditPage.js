import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditPage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.match.params.id, expense);
    this.props.history.push('/');
  };

  onClick = () => {
    this.props.startRemoveExpense(this.props.expense.id);

    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className='page-header'>
          <div className='container'>
            <h2 className='page-header__title'>Edit Expense</h2>
          </div>
        </div>
        <ExpenseForm onSubmit={this.onSubmit} {...this.props.expense} />
        <div className='container'>
          <button className='btn btn--remove' onClick={this.onClick}>
            Remove Expense
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((exp) => exp.id === props.match.params.id),
  };
};

const mapDispatchToProps = (dispatch) => ({
  startEditExpense(id, expense) {
    dispatch(startEditExpense(id, expense));
  },
  startRemoveExpense(id) {
    dispatch(startRemoveExpense({ id }));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPage);
