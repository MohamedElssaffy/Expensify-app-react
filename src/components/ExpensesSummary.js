import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import { Link } from 'react-router-dom';

import selectExpenses from '../selectors/expenses';
import { totalAmount } from '../selectors/expense-total';

export const ExpensesSummary = (props) => (
  <div className='page-header'>
    <div className='container'>
      <h2 className='page-header__title'>
        Viewing <span>{props.expenseCount}</span>
        {props.expenseCount === 1 ? ' expense ' : ' expenses '}
        totalling{' '}
        <span>{numeral(props.expensesTotal / 100).format('$0,0.00')}</span>
      </h2>
      <div className='page-header__actions'>
        <Link to='/create' className='btn'>
          Add Expense
        </Link>
      </div>
    </div>
  </div>
);

const maptStateToProps = (state) => {
  const expenses = selectExpenses(state);
  return {
    expenseCount: expenses.length,
    expensesTotal: totalAmount(expenses),
  };
};

export default connect(maptStateToProps)(ExpensesSummary);
