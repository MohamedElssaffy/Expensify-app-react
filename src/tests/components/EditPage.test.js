import React from 'react';
import { shallow } from 'enzyme';

import { EditPage } from '../../components/EditPage';
import expenses from '../seed-data/expenses';

let startEditExpense, startRemoveExpense, history, match, wrapper;

beforeEach(() => {
  startEditExpense = jest.fn();
  startRemoveExpense = jest.fn();
  history = { push: jest.fn() };
  match = { params: { id: expenses[1].id } };

  wrapper = shallow(
    <EditPage
      expense={expenses[1]}
      history={history}
      startEditExpense={startEditExpense}
      startRemoveExpense={startRemoveExpense}
      match={match}
    />
  );
});

test('Should render Edit page', () => {
  expect(wrapper).toMatchSnapshot();
});

test('Should handle edite expense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startEditExpense).toHaveBeenLastCalledWith(
    match.params.id,
    expenses[1]
  );
});

test('Should handle remove expense', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startRemoveExpense).toHaveBeenLastCalledWith(match.params.id);
});
