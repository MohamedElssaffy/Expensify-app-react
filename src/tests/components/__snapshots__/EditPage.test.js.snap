// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Edit page 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="container"
    >
      <h2
        className="page-header__title"
      >
        Edit Expense
      </h2>
    </div>
  </div>
  <ExpenseForm
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id="2"
    note=""
    onSubmit={[Function]}
  />
  <div
    className="container"
  >
    <button
      className="btn btn--remove"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
